### ISSUE TITLE
Feature request: Add integration to Varnish's Refresh/Purge API

### ISSUE THREAD

**Issue by phg49389** [betheluniversity/cache-clear#2]
With the most recent change to Varnish Prod, it now has an API available that can remotely purge URLs from its cache, Refresh a cached URL, or use regular expressions to remotely add rules to Varnish's Ban List. However, that functionality is locked down both by IP of the client as well as Basic Auth. 

- All requests are sent to `GET https://purge.bethel.edu/`, and the Headers of the request determine the action taken by the API
- Requests must have the correct BasicAuth header to be authorized
- Header `API-Action` is mandatory, and can have value ['purge', 'refresh', 'ban']
- Headers `Purge-Host` and `Purge-URL` are mandatory for both Purging and Refreshing
    - `Purge-Host` is the base of the URL to be purged -- something like www.bethel.edu
    - `Purge-URL` is the actual route to be purged -- something like /undergrad/academics
- Header `Ban-Expression` is mandatory for Banning
    - `Ban-Expression` should use [this syntax](https://varnish-cache.org/docs/4.1/reference/varnish-cli.html#ban-expressions) since it will be passed through to the backend of Varnish directly.
- Sanitizing all inputs must be done in this module, since there's no way to sanitize on Varnish's side. We'll need to check at least the following things:
    - The host being Purged/Refreshed/Banned is even being cached by Varnish (Maybe store a list in config.py of these hosts?)
    - The full path (host + url) of Purge/Refresh exists (Maybe send a request to that path before sending the API call?)
    - The ban expression is syntactically valid (might be able to be done with a regex?)

I can provide more details on how to use the Refresh/Purge API when this feature starts getting developed. 

A simple interface that may suffice to start out with would be to have a Text Field that people paste a URL they want to purge or refresh, a radio button to select either Purge or Refresh, and a submit button. The Python code could take the url, split it according to something like `(https?://)?([^/]+?)(.*)`, and use the 2nd group as the Purge-Host and the 3rd group as Purge-URL. From there sanitization would proceed.

Then for Banning, there could be a second Text Field and a button labeled "Ban". On submission, the text would be syntax-checked, and then if they specify `req.http.host` anywhere inside that ban expression, we would have to use the value they passed as a regex and see if it matched at least one of the valid hosts in config.py.

Once we've done all the input sanitizing, then it should be a pretty trivial matter to construct the API request, send it, and make sure it gets a 200 response back from Varnish.

**Comment by c-schwarze** [comment]
As we might be moving towards NOT cacheing images via varnish, we probably won't have to do this.

**Comment by phg49389** [comment]
Another idea that I had for the Ban interface would be to keep the two Textfields for Host and URL, like Purge/Refresh, but if the Ban button was clicked it would automatically send something like this:

```
Ban-Expression: "req.http.host ~ %s && req.url ~ %s" % (Host.value, URL.value)
```

By using that functionality, it would make the conversion to using "[Smart Bans](https://book.varnish-software.com/3.0/Cache_invalidation.html#smart-bans)" a lot more simple. Here's a [Confluence article](https://confluence.bethel.edu/x/HAJEAg) I wrote up about how to enable Smart Bans in our Varnish. After Smart Bans were enabled on Varnish's side, then the above line would change to this:

```
Ban-Expression: "obj.http.x-host ~ %s && obj.http.x-url ~ %s" % (Host.value, URL.value)
```

Lastly, it might be useful to keep a misc. text field for Banning, so that we can do custom expressions other than banning by host/url just like purges & refreshes. It would still suffer from no syntax checking, but it would be a backup option instead of the primary interface for banning.

**Comment by phg49389** [comment]
The regex for Ban expression syntax is a bit of a monster... 
```
((req|obj)\.(url|status|http\.[A-Za-z0-9-]+)\s?(!?~|[!=]=)\s?([^ \n]+)(\s?&&\s?(?=(req|obj)\.(url|status|http\.[A-Za-z0-9-]+)\s?(!?~|[!=]=)\s?([^ \n]+)))?)+
```

1. Assert the first operand is a commonly accepted object to compare against in VCL
2. Assert that the second operand is one of the valid operators
3. Grab as many characters as it can before encountering a space or newline to parse the third operand -- the value to be compared
4. Once it has parsed an entire comparison of three operands, it allows for an optional && at the end on the condition that there's another comparison after the &&
5. Finally, assert that it has to match at least one comparison in order to produce a match, but it is designed to catch however many comparisons get chained together (assuming that the chain doesn't end with another &&).

I consulted the C source code of Varnish to see if they have a more elegant way to do it, but they parse the Ban expression like a code interpreter instead of like a program parsing a string. If someone wants to take a stab at replicating the C code in Python, here are the two source files that I found:

- https://github.com/varnishcache/varnish-cache/blob/d001cdd2045aa484a968d09859cddf59dee372cb/bin/varnishd/cache/cache_vrt.c#L681
- https://github.com/varnishcache/varnish-cache/blob/e1d47464de79fac1823688b7437d5c8dccf27528/bin/varnishd/cache/cache_ban_build.c#L218
